// graphql-yoga/esm/index.js
import yargs from 'yargs';
import { hideBin } from 'yargs/helpers';
import { loadConfig } from 'graphql-config';
import { CodeFileLoader } from '@graphql-tools/code-file-loader';
import { createServer } from '@graphql-yoga/node';
import { addMocksToSchema } from '@graphql-tools/mock';
const terminateEvents = ['SIGINT', 'SIGTERM'];
function registerTerminateHandler(callback) {
    for (const eventName of terminateEvents) {
        process.on(eventName, () => callback(eventName));
    }
}
export const YogaExtensions = (api) => {
    const codeFileLoader = new CodeFileLoader({
        noPluck: true,
    });
    api.loaders.schema.register(codeFileLoader);
    api.loaders.documents.register(codeFileLoader);
    return {
        name: 'Yoga',
    };
};
export function graphqlYoga() {
    return yargs(hideBin(process.argv)).command('$0', 'Serves GraphQL over HTTP using your GraphQL Config', (builder) => {
        builder.option('project', {
            type: 'string',
            description: 'Project name',
        });
        builder.option('mock', {
            type: 'boolean',
            description: 'Mock the given schema',
        });
    }, async ({ project = 'default', mock }) => {
        console.info(`Loading GraphQL Config from ${process.cwd()}`);
        const config = await loadConfig({
            extensions: [YogaExtensions],
            throwOnMissing: false,
            throwOnEmpty: false,
        });
        console.log(`Loading project: ${project}`);
        const projectConfig = config === null || config === void 0 ? void 0 : config.getProject(project);
        console.log(`Loading GraphQL Schema of ${project}`);
        let schema = await (projectConfig === null || projectConfig === void 0 ? void 0 : projectConfig.getSchema());
        if (!schema) {
            console.warn(`Could not find schema for project ${project} fallback to default schema`);
        }
        if (mock) {
            if (!schema) {
                console.warn('No schema found for mocking. Skipping mocking.');
            }
            else {
                console.log(`Adding mocks to the schema`);
                schema = addMocksToSchema({ schema });
            }
        }
        console.log(`Building GraphQL Server`);
        const graphQLServer = createServer({
            schema,
        });
        console.log(`Starting GraphQL Server`);
        await graphQLServer.start();
        registerTerminateHandler(() => {
            graphQLServer.stop();
        });
    }).argv;
}
export * from '@graphql-yoga/node';
